File: frontend/src/features/admin/types/company.ts
/**
 * 업체 상태
 */
export type CompanyStatus = "approved" | "pending" | "rejected" | "suspended";

/**
 * 계약 상태
 */
export type ContractStatus = "active" | "pending" | "expired";

/**
 * 업체 정보
 */
export interface Company {
  id: string;
  name: string; // 업체명
  businessNumber: string; // 사업자번호
  representativeName: string; // 대표자명
  email: string;
  phone: string;
  address: string;
  region: string; // 지역 (서울, 경기 등)
  status: CompanyStatus;
  approvedAt?: string | null; // 승인일 (ISO)
  rejectedReason?: string | null; // 반려 사유
  deviceCount: number; // 관리 중인 장치 수
  customerCount: number; // 담당 고객(대상자) 수
  createdAt: string;
  updatedAt: string;
}

/**
 * 업체 상세 정보
 */
export interface CompanyDetail extends Company {
  serviceHistory: ServiceRecord[]; // 서비스 이력
  managedDevices: ManagedDevice[]; // 관리 중인 장치 목록
  customers: Customer[]; // 담당 고객 목록
}

/**
 * 서비스 이력
 */
export interface ServiceRecord {
  id: string;
  type: "installation" | "repair" | "inspection" | "replacement";
  subjectId: string;
  subjectName: string;
  deviceId: string;
  deviceSerialNumber: string;
  description: string;
  performedAt: string; // ISO 날짜시간
  performedBy: string; // 담당자 이름
  cost?: number; // 비용 (옵션)
}

/**
 * 관리 중인 장치 (업체 상세용)
 */
export interface ManagedDevice {
  id: string;
  serialNumber: string;
  modelName: string;
  status: string;
  assignedSubjectName?: string;
}

/**
 * 담당 고객 (업체 상세용)
 */
export interface Customer {
  id: string;
  name: string;
  phone: string;
  deviceSerialNumber?: string;
  lastServiceDate?: string; // ISO 날짜
}

/**
 * 업체 목록 조회 응답
 */
export interface CompanyListResponse {
  success: boolean;
  totalCount: number;
  companies: Company[];
}

/**
 * 업체 상세 조회 응답
 */
export interface CompanyDetailResponse {
  success: boolean;
  company: CompanyDetail;
}

/**
 * 업체 등록 요청
 */
export interface CreateCompanyRequest {
  name: string;
  businessNumber: string;
  representativeName: string;
  email: string;
  phone: string;
  address: string;
  region: string;
  certificationValidUntil: string; // 인증 유효 기간 (ISO 날짜)
  contractStatus: ContractStatus; // 계약 상태
  businessRegistrationNumber: string; // 사업자등록증 번호
  sealInfo: string; // 직인 정보
  initialAccountId: string; // 초기 계정 ID
  initialPassword: string; // 초기 비밀번호
}

/**
 * 업체 승인/반려 요청
 */
export interface ApproveCompanyRequest {
  companyId: string;
  status: "approved" | "rejected";
  rejectedReason?: string; // 반려 시 필수
}

/**
 * 업체 수정 요청
 */
export interface UpdateCompanyRequest {
  name?: string;
  email?: string;
  phone?: string;
  address?: string;
  region?: string;
  status?: CompanyStatus;
}

File: backend/company-service/src/main/java/com/ddp/company/entity/Company.java
package com.ddp.company.entity;

import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

// 업체 엔티티 클래스
@Entity
@Table(name = "companies")
@EntityListeners(AuditingEntityListener.class)
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Builder
@Getter
@Setter
public class Company {

    // 업체 고유 ID (기본키)
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // 업체명
    @Column(nullable = false)
    private String name;

    // 사업자등록번호
    @Column(nullable = false, unique = true)
    private String businessNumber;

    // 대표자명
    @Column(nullable = false)
    private String representativeName;

    // 이메일 주소
    @Column(nullable = false, unique = true)
    private String email;

    // 전화번호
    @Column(nullable = false)
    private String phone;

    // 주소
    @Column(nullable = false)
    private String address;

    // 지역 (서울, 경기 등)
    @Column(nullable = false)
    private String region;

    // 위도
    @Column
    private Double latitude;

    // 경도
    @Column
    private Double longitude;

    // 영업시간
    @Column
    private String businessHours;

    // 업체 소개
    @Column(columnDefinition = "TEXT")
    private String description;

    // 평점 (0-5)
    @Column
    @Builder.Default
    private Double rating = 0.0;

    // 리뷰 수
    @Column
    @Builder.Default
    private Integer reviewCount = 0;

    // 업체 상태
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private CompanyStatus status;

    // 승인일
    @Column
    private LocalDateTime approvedAt;

    // 반려 사유
    @Column
    private String rejectedReason;

    // 초기 계정 ID
    @Column
    private String initialAccountId;

    // 관리 중인 장치 수
    @Column
    @Builder.Default
    private Integer deviceCount = 0;

    // 담당 고객(대상자) 수
    @Column
    @Builder.Default
    private Integer customerCount = 0;

    // 생성 시간
    @CreatedDate
    @Column(updatable = false)
    private LocalDateTime createdAt;

    // 수정 시간
    @LastModifiedDate
    private LocalDateTime updatedAt;

    // 삭제 시간 (Soft Delete)
    @Column
    private LocalDateTime deletedAt;

    // 삭제자 ID (관리자 ID)
    @Column
    private Long deletedBy;

    // 서비스 이력 목록 (1:N 관계)
    @OneToMany(mappedBy = "company", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<ServiceRecord> serviceHistory = new ArrayList<>();

    // 관리 중인 장치 목록 (1:N 관계)
    @OneToMany(mappedBy = "company", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<ManagedDevice> managedDevices = new ArrayList<>();

    // 담당 고객 목록 (1:N 관계)
    @OneToMany(mappedBy = "company", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<Customer> customers = new ArrayList<>();

    /**
     * 업체 소프트 삭제 수행
     */
    public void softDelete(Long adminId) {
        this.deletedAt = LocalDateTime.now();
        this.deletedBy = adminId;
        this.status = CompanyStatus.REJECTED; // 삭제된 업체는 REJECTED 상태로 변경
    }

    /**
     * 삭제 여부 확인
     */
    public boolean isDeleted() {
        return this.deletedAt != null;
    }
}
